contracts/eosio.system/src/peer_keys.cpp

```diff
@@ -0,0 +1,38 @@
+#include <eosio.system/eosio.system.hpp>
+
+#include <eosio/eosio.hpp>
+
+namespace eosiosystem {
+
+   void system_contract::regpeerkey( const name& proposer_finalizer_name, const public_key& key ) {
+      require_auth(proposer_finalizer_name);
+      check(!std::holds_alternative<eosio::webauthn_public_key>(key), "webauthn keys not allowed in regpeerkey action");
+
```

Comment:
Seems like we should verify that `proposer_finalizer_name` is in the producer table.

Reply:
> Seems like we should verify that `proposer_finalizer_name` is in the producer table

Seems reasonable



> Actually, I think it would be good to verify it is in the producer table and is in the top 100 votes.

Why?

Reply:
> Why?

It would prevent just anyone from adding a peer key in attempt to gain access to the internal gossip bp messages which report server addresses.

Reply:
I don't think that verifying that proposer_finalizer_name is in the producer table is solving an existing issue, since peer gossip messages would not be sent to nodes that are not in the producer schedule anyways (the fact that a key is populated on-chain is a necessary, but not sufficient condition). 
And since populating extra keys on-chain is not a probable attack vector, I feel that it would just add unnecessary code, although I don't mind adding it if you guys think otherwise.

Reply:
Hum, I must have misunderstood your comment [here](https://github.com/AntelopeIO/spring/pull/1265#issuecomment-2754870181)?

Reply:
A node has to be run with the private key of the public key registered via `regpeerkey`. If that node then connects to any bp gossip peer it will receive the complete list of bp gossip peers.

Reply:
That was not my understanding, I thought that `regpeerkey` was only used to validate that the gossiped address was correct, not as a limit to the gossip network.

I think that such a control should be implemented in `handle_message( gossip_bp_peers_message& msg )`, more precisely `validate_gossip_bp_peers_message(msg)`, because this can use the current list of top producers, instead of what the list was when `regpeerkey` was called, which could be a long time ago.

Reply:
That would require a cached list of top producers similar to the cached list of peer keys.

Reply:
> > Why?
> 
> It would prevent just anyone from adding a peer key in attempt to gain access to the internal gossip bp messages which report server addresses.

I'm doubtful requiring an account to be in the top 100 is a viable protection against that -- an account could just borrow some EOS for a small period of time to vote itself up; could it even do it via flash loan?

> That would require a cached list of top producers similar to the cached list of peer keys.

Or the onus could be on the contract to only expose the keys for the current producers.

Reply:
> That would require a cached list of top producers similar to the cached list of peer keys.

Maybe we should just worry about the top 21 in the schedule and limit gossipping to these. When a new node enters the schedule, we can initiate a connection to it if it has registered a peer key.

Reply:
I don't follow how that would work. If you don't have the connection information for the new BP that came into the schedule how do you initiate a connection to it?

Reply:
Maybe we are talking past each other. My understanding is you are proposing to only gossip the active/pending top 21 BPs. That means that any given node only knows the connection information of the top 21. When a new BP comes into the schedule, you know that peer key and account name from the on-chain contract. How do you know the server address to connect to at that point?

Currently the PR to Spring gossips all validated bp connections to all other validated bp connections for any bp in the `peerkeys` table. Therefore, all bp peers know the addresses of all the others. It then connects/disconnects from them as they come into and out of the schedule.

Reply:
Yes, sorry,  I realized too late my previous comment was wrong.

Maybe it is up to the new producer to send an initial peer gossip request to another producer (a configured one), which should respond with the list of all producers and their addresses, and then the new producer connects to all its BP peers.
Still limiting gossiping in net_plugin to the current top-21 producers.

Reply:
Ok, yes, that scheme would work. Maybe each node could maintain a validated set of bp peers that it knows about but only ever gossip out top 21. This requires a lot more gossip messages (every time the schedule changes). Currently gossip message only sent as any new connection info is acquired.

Reply:
> > Maybe each node could maintain a validated set of bp peers that it knows about
> 
> Could they do that if the gossip connections are only between the top 21?

You have to have a bootstrap connection into the internal bp network. That is true no matter what scheme we use. So one peer (and maybe only one which makes it a bit fragile) has to know about the connection.